### PR TITLE
Hierarchy expansion on search

### PR DESCRIPTION
**Pull request by igor-dv**
## What I did

1. Expand all hierarchy when in a filter mode
2. Highlight found patterns
3. Added debouncing to the filter

filter-expand-highlight

## How to test
run `cra-kitchen-sink` and perform filter
 
## TBD
Add tests !!

### CODE REVIEW

**Comment by shilman** [comment]
dude you're on 🔥 !! looking incredible.

File: lib/ui/src/modules/ui/components/left_panel/text_filter.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import PropTypes from 'prop-types';
 import React from 'react';
-import debounce from 'lodash.debounce';
+import { debounce } from 'lodash';
```
Comment (Hypnosphi) [inline lib/ui/src/modules/ui/components/left_panel/text_filter.js]: What was wrong with the standalone package?

Comment (igor-dv) [inline lib/ui/src/modules/ui/components/left_panel/text_filter.js]: I didn't manage to mock it...

Comment (Hypnosphi) [inline lib/ui/src/modules/ui/components/left_panel/text_filter.js]: I think it can be mocked just as identity function: fn => fn

Comment (igor-dv) [inline lib/ui/src/modules/ui/components/left_panel/text_filter.js]: Yeah, that what I did.. But problem looks like in a way of the import.. In the transpiled version 
 of the `import ... from lodash.debounce` I am getting a reference to the `debounce` function. And in transpiled version of named import - the reference is to the `lodash` object, in this case I managed to replace its method. :face_with_head_bandage:

**Comment by ndelangen** [comment]
🔥

File: lib/ui/src/modules/ui/components/left_panel/stories_tree/tree_decorators.js [hunk 2]
```diff
@@ -84,22 +84,60 @@ ContainerDecorator.propTypes = {
...
-    };
+class HeaderDecorator extends React.Component {
+  decorateNameMatchedToSearchTerm(node, style) {
+    const { storyFilter, name } = node;
+
+    if (!storyFilter) {
+      return name;
+    }
+
+    const filterRegex = new RegExp(`(${storyFilter})`, 'i');
```
Comment (Hypnosphi) [inline lib/ui/src/modules/ui/components/left_panel/stories_tree/tree_decorators.js]: Looks like this will break if with special characters. I'd suggest to use [Fuse.js](fusejs.io) with `{ includeMatches: true }` in `libs/filters.js` and store the matches information in some spesial field on the filtered stories

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

**Comment by igor-dv** [comment]
Chatted with @Hypnosphi  in slack. 

Our current `fuzzysearch` lib is a bit limiting, since the only result it gives is whether the string matches to the query.

Already mentioned [Fuse.js](https://github.com/krisk/Fuse) lib is very promising and can help us improve both filtering and highlighting features (built-in matches indices in the result). However they have a [PR](https://github.com/krisk/Fuse/pull/167) that is vital for us. So we can make a fork (like with the react-treabeard).

Other alternative might be this small lib - https://github.com/mattyork/fuzzy/blob/master/lib/fuzzy.js.
It **has no** built in matches feature, but can be easily used as a source of truth for both - filtering and highlighting.

In other words I think this change is bigger than just PR to master.. 

What do you guys think ?

**Comment by Hypnosphi** [comment]
> However they have a PR that is vital for us. So we can make a fork (like with the react-treabeard).

We can make a fork and send a new PR with tests =)

> It has no built in matches feature, but can be easily used as a source of truth for both - filtering and highlighting.

How exactly would you use it for highlighting?

**Comment by shilman** [comment]
@igor-dv I think the current behavior is perfect for most real-world component libraries, which never grow particularly big/complex. You're always welcome to make it better, but if it was me I would just ship it as is.

**Review by shilman — APPROVED** [review]
LATM! (A = awesome!)

**Comment by Hypnosphi** [comment]
@shilman please type '\` in the filter and look into console

**Comment by igor-dv** [comment]
I've fixed the special characters issue.

@shilman regarding the highlighting, here is an example when it breaks..  

image

I think if there are no strong objections we can merge this one (after I'll add few tests, probably tomorrow) And add the Fuse.js as a  separate PR (and maybe as part of the next release).

**Comment by Hypnosphi** [comment]
Great! The first issue is gone, the second one can be solved in a scope of another PR

**Review by Hypnosphi — APPROVED** [review]
